lib/rules/jsx-no-useless-fragment.js

```diff
@@ -0,0 +1,198 @@
+}
+
+/**
+ * @param {string} text
+ * @returns {string}
+ */
+function trimLikeReact(text) {
+  const leadingSpaces = /^\s*/.exec(text)[0];
+  const trailingSpaces = /\s*$/.exec(text)[0];
+
+  const start = leadingSpaces.includes('\n') ? leadingSpaces.length : 0;
+  const end = trailingSpaces.includes('\n') ? text.length - trailingSpaces.length : text.length;
+
+  return text.slice(start, end);
+}
+
+/**
+ * Test if node is like `<Fragment key={_}>_</Fragment>`
+ * @param {JSXElement} node
+ * @returns {boolean}
+ */
+function isKeyedElement(node) {
+  return node.type === 'JSXElement' &&
+    node.openingElement.attributes &&
+    node.openingElement.attributes.some(attribute => (
+      attribute.type === 'JSXAttribute' &&
+      attribute.name &&
+      attribute.name.type === 'JSXIdentifier' &&
+      attribute.name.name === 'key'
+    ));
+}
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    fixable: 'code',
+    docs: {
+      description: 'Disallow unnecessary fragments',
+      category: 'Possible Errors',
+      recommended: false,
+      url: docsUrl('jsx-no-useless-fragment')
+    },
+    messages: {
+      NeedsMoreChidren: 'Fragments should contain more than one child - otherwise, there‘s no need for a Fragment at all.',
+      ChildOfHtmlElement: 'Passing a fragment to an HTML element is useless.'
+    }
+  },
+
+  create(context) {
+    const reactPragma = pragmaUtil.getFromContext(context);
+    const fragmentPragma = pragmaUtil.getFragmentFromContext(context);
+
+    /**
+     * Test whether a node is an padding spaces trimmed by react runtime.
+     * @param {ASTNode} node
+     * @returns {boolean}
+     */
+    function isPaddingSpaces(node) {
+      return isJSXText(node) &&
+        isOnlyWhitespace(node.raw) &&
+        node.raw.includes('\n');
+    }
+
+    /**
+     * Test whether a JSXElement has less than two children, excluding paddings spaces.
+     * @param {JSXElement|JSXFragment} node
+     */
+    function hasLessThanTwoChildren(node) {
+      if (!node || !node.children || node.children.length < 2) {
+        return true;
+      }
+
+      return (
+        node.children.length -
+        (+isPaddingSpaces(node.children[0])) -
+        (+isPaddingSpaces(node.children[node.children.length - 1]))
+      ) < 2;
+    }
+
+    /**
+     * @param {JSXElement|JSXFragment} node
+     * @returns {boolean}
+     */
+    function isChildOfHtmlElement(node) {
+      return node.parent.type === 'JSXElement' &&
+        node.parent.openingElement.name.type === 'JSXIdentifier' &&
+        /^[a-z]+$/.test(node.parent.openingElement.name.name);
+    }
+
+    /**
+     * @param {ASTNode} node
+     * @returns {boolean}
+     */
+    function canFix(node) {
+      // Fragments that are child elements can always be fixed
+      if (!(node.parent.type === 'JSXElement' || node.parent.type === 'JSXFragment')) {
+        // const a = <></>
+        if (node.children.length === 0) {
```

Comment:
this is covered implicitly by `[].some(…)` always returning false, so this block can be removed

Reply:
Actually this cannot be removed otherwise tests fails. When `node.children` is an empty array, current behavior returns false. With the block removed, it returns true.

Reply:
That doesn't make sense; `[].some(() => true)` returns `false`.

Reply:
I confirmed again that test fails when the if statement is removed. https://travis-ci.org/golopot/eslint-plugin-react/jobs/582110265

Reply:
ohhh i get what i'm confused about. the `.some` will return false, preventing the early exit, and then will fall through to the `return true`. sorry for the noise :-)